pczt/src/transparent.rs

```diff
@@ -0,0 +1,411 @@
+    ///
+    /// - These may be used by the Signer to inspect parts of `script_pubkey` or
+    ///   `redeem_script`.
+    pub(crate) ripemd160_preimages: BTreeMap<[u8; 20], Vec<u8>>,
+
+    /// Mappings of the form `key = SHA256(value)`.
+    ///
+    /// - These may be used by the Signer to inspect parts of `script_pubkey` or
+    ///   `redeem_script`.
+    pub(crate) sha256_preimages: BTreeMap<[u8; 32], Vec<u8>>,
+
+    /// Mappings of the form `key = RIPEMD160(SHA256(value))`.
+    ///
+    /// - These may be used by the Signer to inspect parts of `script_pubkey` or
+    ///   `redeem_script`.
+    pub(crate) hash160_preimages: BTreeMap<[u8; 20], Vec<u8>>,
+
+    /// Mappings of the form `key = SHA256(SHA256(value))`.
+    ///
+    /// - These may be used by the Signer to inspect parts of `script_pubkey` or
+    ///   `redeem_script`.
+    pub(crate) hash256_preimages: BTreeMap<[u8; 32], Vec<u8>>,
+
+    /// Proprietary fields related to the note being spent.
+    #[getset(get = "pub")]
+    pub(crate) proprietary: BTreeMap<String, Vec<u8>>,
+}
+
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize, Getters)]
+pub struct Output {
+    //
+    // Transparent effecting data.
+    //
+    // These are required fields that are part of the final transaction, and are filled in
+    // by the Constructor when adding an output.
+    //
+    pub(crate) value: u64,
+    pub(crate) script_pubkey: Vec<u8>,
+
+    /// The script required to spend this output, if it is P2SH.
+    ///
+    /// Set to `None` if this is a P2PKH output.
+    pub(crate) redeem_script: Option<Vec<u8>>,
+
+    /// A map from a pubkey to the BIP 32 derivation path at which its corresponding
+    /// spending key can be found.
+    ///
+    /// - The pubkeys should appear in `script_pubkey` or `redeem_script`.
+    /// - Each entry is set by an Updater.
+    /// - Individual entries may be required by a Signer.
+    /// - It is not required that the map include entries for all of the used pubkeys.
+    ///   In particular, it is not possible to include entries for non-BIP-32 pubkeys.
+    #[serde_as(as = "BTreeMap<[_; 33], _>")]
+    pub(crate) bip32_derivation: BTreeMap<[u8; 33], Zip32Derivation>,
+
+    /// Proprietary fields related to the note being spent.
+    #[getset(get = "pub")]
+    pub(crate) proprietary: BTreeMap<String, Vec<u8>>,
+}
+
+impl Bundle {
+    /// Merges this bundle with another.
+    ///
+    /// Returns `None` if the bundles have conflicting data.
+    pub(crate) fn merge(mut self, other: Self) -> Option<Self> {
+        // Destructure `other` to ensure we handle everything.
+        let Self {
+            mut inputs,
+            mut outputs,
+        } = other;
+
+        // If the other bundle has more inputs or outputs than us, move them over; these
+        // cannot conflict by construction.
+        self.inputs.extend(inputs.drain(self.inputs.len()..));
+        self.outputs.extend(outputs.drain(self.outputs.len()..));
```

Comment:
This does not check `tx_modifiable`. Is it supposed to?

Reply:
Good catch, it should enforce that if `tx_modifiable` on a bundle indicates no modifications, then merging of the other bundle should fail if it would add inputs or outputs (a combine cannot alter or remove inputs or outputs).